src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -397,7 +455,14 @@ protected void setHeads(@Nonnull Collection<FlowNode> heads) {
             currentParallelStartNode = currentParallelStart.forkStart;
             myCurrent = currentParallelStart.unvisited.pop();
             myNext = myCurrent;
-            nextType = NodeType.PARALLEL_BRANCH_END;
+
+            // We may have a start type, so we need to override the beginning type
+            NodeType tempType = getNodeType(myCurrent);
```

Comment:
OK, so you added a nice new method to determine what type the node is, just to manipulate the result and almost not care what it gives you anyway?
Can't say it's wrong just that something smells funny.